[mlir][emitc] Add 'emitc.while' and 'emitc.do' ops to the dialect

Vladislave0-0 opened the pull request:
This MR adds:
- `emitc::WhileOp` and `emitc::DoOp` to the EmitC dialect (аfter all the discussions, only `emitc::DoOp` )
- Emission of the corresponding ops in the CppEmitter
- Conversion from the SCF dialect to the EmitC dialect for the ops
- Corresponding tests

Vladislave0-0 committed:
[mlir][emitc] Add 'emitc.while' and 'emitc.do' ops to the dialect

This MR adds:
- 'emitc::WhileOp' and 'emitc::DoOp' to the EmitC dialect
- Emission of the corresponding ops in the CppEmitter
- Conversion from the SCF dialect to the EmitC dialect for the ops
- Corresponding tests

jacquesguan commented on mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:
```diff
@@ -1572,4 +1572,156 @@ def EmitC_SwitchOp : EmitC_Op<"switch", [RecursiveMemoryEffects,
+  let arguments = (ins);
+  let results = (outs); 
+  let regions = (region MaxSizedRegion<1>:$conditionRegion,
+                        MaxSizedRegion<1>:$bodyRegion);
```
`SizedRegion`

jacquesguan commented on mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:
```diff
@@ -1572,4 +1572,156 @@ def EmitC_SwitchOp : EmitC_Op<"switch", [RecursiveMemoryEffects,
+
+  let arguments = (ins);
+  let results = (outs); 
+  let regions = (region MaxSizedRegion<1>:$conditionRegion,
```
Should we put `condition` to arguments? In CPP, the condition should be an expression (or simple declaration).

jacquesguan replied:
I think maybe keep it similar to 'emitc.if'. No need to use a region for condition. 
Something like ` let arguments = (ins I1:$condition);` is fine.

kchibisov replied:
No, this won't work, since you need to `load`, otherwise what would be the point of such `while` and `do`. If we go with `inline`, you'd also must ensure then that it always happens into the `while` cond and at this point, it seems like simpler to just put the `cond` into it.

`if` on the otherside doesn't have a problem of `load`-ing, etc and also doesn't care about inline, since it generally.

You also won't be able to translate to such `emitc.while` from `scf.while` into anything other than `exit = false; do { cond = eval(); } while (cond)`.

Don't get me wrong, we've tried using plain `cond`, but the fact that you must have `load` working and that you must have `scf` translation, so operation is not dead made us use the current approach.

One could rely on `emitc.expression` being inlined, but it's much less clear from the user perspective, since you generally pass the result, not the entire expression in such cases.

kchibisov replied:
How do you propose to deal with no-inline `C` expressions that have loads and ensure translation correctness.

We can not pass `emitc.expression` as an argument, we can only pass `%1` as an argument that may be backed by `emitc.expression`, which may get inlined based on some conditions. That's a lot of `may`, while the proposed approach looks like `scf` and uses the same logic with `emitc.expression`, but is much more clear in my opinion, since you know exactly how it'll end up in the users code.

Unless you can pass `emitc.experssion` as an argument itself, I don't see how to do so.

kchibisov replied:
I'd add for clarification, since I've been helping @Vladislave0-0 to make it up and did some internal reviews before sending upstream, I'd also speak on his behalf on things we've agreed upon sending.

kchibisov replied:
The problem is that you can not use `emitc.expression` as an operand because it's not a type, so you can only pass a `bool` pretty much. 

>OTOH, if the condition's emitc.expression is located outside the loop then inlining that expression would change program semantics from a single to multiple evaluations of the expression.

This will make `while` completely useless, it must always be inlined for `load` to be a part of the condition and be inside the cond  `while (cond)`.

>How about requiring the condition region of these loop ops to contain a single emitc.expression op feeding an emitc.yield? That should both utilize emitc.expression and avoid any inlining issues.

Do you want to nest it one more time? That should work, a bit ugly, but I guess it's fine.

aniragil replied:
>Just to make it clear, to do you want to have it that way
>```
>emitc.while {
>    emitc.expression : i1 {
>    }
>}
>```
This loop doesn't seem to have a condition region. I meant to have a condition region as the patch suggests but let it hold a single `emitc.expression` whose value is then yielded, i.e.:
```
emitc.while {
    // Condition region
    %e = emitc.expression : i1 {
       %r = ... : i1
       yield %r : i1
    }
    yield %e : i1
} do {
   // Body region
}
```
Then  `CppEmitter::emitWhile()` can call `CppEmitter::emitExpression()` on the expression instead of replicating its code.
Using the dialect's operation that models expressions would capture the exact semantics in the IR, let `form-expressions` merge feeding expressions into the condition expression and simplify verification and translation code.

kchibisov replied:
yes, that what I've implied when I wrote that, it's just @Vladislave0-0 is a bit busy now, so implementation stalled a bit, but we'll adopt to the suggested changes.

Vladislave0-0 replied:
Hello, @aniragil! Thank you. I rewrote the condition region structure according to the discussion above.

Vladislave0-0 commented:
While working on the `emitc::WhileOp`, a bug was discovered in the `emitc::ExpressionOp`.

#143894 - This MR fixes it. Check out this MR first, please.

Vladislave0-0 committed:
[mlir][emitc] Change conditional region structure

Change the canonical structure of a conditional region:
- The condition region must contain exactly one block with:
  1. An `emitc.expression` operation producing an i1 value
  2. An `emitc.yield` passing through the expression result
- The body region must not yield any values

Vladislave0-0 commented:
@aniragil, gentle ping. I have resolved all what you requested

aniragil reviewed (changes requested):
> @aniragil, gentle ping. I have resolved all what you requested

Thanks, taking a look. Some initial comments inline.

Vladislave0-0 commented on mlir/lib/Conversion/SCFToEmitC/SCFToEmitC.cpp:
```diff
@@ -332,11 +332,260 @@ LogicalResult IndexSwitchOpLowering::matchAndRewrite(
+    auto exprOp = rewriter.create<emitc::ExpressionOp>(loc, TypeRange{i1Type});
+    Region &exprRegion = exprOp.getBodyRegion();
+
+    rewriter.inlineRegionBefore(whileOp.getBefore(), exprRegion,
```
> The `before` block is not guaranteed to contain a valid C expression, e.g.:
> 
> ```mlir
> func.func @double_use(%p : !emitc.ptr<i32>) -> i32 {
>   %init = emitc.literal "1.0" : i32
>   %var  = emitc.literal "1.0" : i32
>   %exit = emitc.literal "10.0" : i32
>   %res = scf.while (%arg1 = %init) : (i32) -> i32 {
>     %used_twice = emitc.call @payload_with_side_effect(%arg1, %p) : (i32, !emitc.ptr<i32>) -> i32
>     %prod = emitc.add %used_twice, %used_twice : (i32, i32) -> i32
>     %sum = emitc.add %arg1, %prod : (i32, i32) -> i32
>     %condition = emitc.cmp lt, %sum, %exit : (i32, i32) -> i1
>     scf.condition(%condition) %arg1 : i32
>   } do {
>   ^bb0(%arg2: i32):
>     %next_arg1 = emitc.call @payload_do(%arg2) : (i32) -> i32
>     scf.yield %next_arg1 : i32
>   }
>   return %res : i32
> }
> ```
> 
> It may also include ops not related to the condition at all. The `emitc.expression` op can in principle be extended to support such sequences using the comma operator, but it currently doesn't (I'm also not sure it'd be very aesthetic)

Thank you for noticing this bug. Given that we can not determine whether it'll really translate to `CExpression` that won't have side effects or translate at all, what do you think about always translating to `do-while` style loop as we do now when handling `scf.while`'s `do-while` style condition, 
```
bool isDoOp = !llvm::equal(beforeBlock.getArguments(), condOp.getArgs());

LogicalResult result =
    isDoOp ? lowerDoWhile(whileOp, variables, context, rewriter, loc)
           : lowerWhile(whileOp, variables, context, rewriter, loc);
```
where such issues are eliminated due to translating into `while` body, which is not required to be a `CExpression`? That way the translation of the `scf.while` itself will always work and it'll behave pretty much the same as it does right now.

As a downside, the new `emitc.while` will be "dead", since it's not part of the translation from the higher dialects, but tbf we can determine after `form-expression` pass, which `emitc.do` can be converted to `while` operation, since we have a very specific pattern of `do { cond_res = cond; if (cond_res) while (cond_res) }`. And do that as a separate optimization.

aniragil replied:
>I wonder what would be the correct way for those things to get translated, given that it may contain operations that can translate to CExpression, but at the time of the transformation, are not CExpression yet, like if you have arith, etc. I thought, that you just translate it anyway, and then it may fail later on due to incomplete translations. Like it's not like emitc.while won't fail now when translating to cpp, since it can not be emitted at all.

Sorry, I wasn't clear: I didn't mean ops of other dialects like `arith`, but `emitc` ops that do not form a single expression (at least not without using C's `comma` operator). For instance, in the following code:
```C
while (true) {
   a[i+1] = b[i*7] + 3;
   foo(t[i] / 5);
   int c = k[i + 11];
   if (c)
     break;
   // do some more work with loads, stores, calls etc.
}
```
the "before" section includes computations unrelated to the condition, so it's not a "classic" while loop where the exit condition is checked before any computation. Since the `while` condition clause requires a single expression, putting all this code there requires turning it into a single expression. This can be done using the comma operator (which `emitc` currently doesn't support), i.e.
```C
while (a[i+1] = b[i*7] + 3, foo(t[i] / 5), k[i + 11]) {
   // do some more work with loads, stores, calls etc.
}
```
But that's far less clear than the original structure IMO.

> what do you think about always translating to do-while style loop ... where such issues are eliminated due to translating into while body, which is not required to be a CExpression? That way the translation of the scf.while itself will always work and it'll behave pretty much the same as it does right now. As a downside, the new emitc.while will be "dead", since it's not part of the translation from the higher dialects, but tbf we can determine after form-expression pass, which emitc.do can be converted to while operation, since we have a very specific pattern of do { cond_res = cond; if (cond_res) while (cond_res) }. And do that as a separate optimization.

Agreed, best to start with a simple and robust lowering. We shouldn't have dead ops, so I'd do this in stages - first a patch to introduce a single loop op to lower `scf.while` in a unified manner. I'm OK with starting either with `emitc.do` or `emitc.while` (with the condition variable initialized to `1`). In any case, lowering can create in the condition region a simple `emitc.expression` that only loads from the condition variable.
The form-expressions pass (if executed) should indeed fold the computation of the condition in the loop body.
A second patch can then introduce a new pass to optimize the loops in both directions, i.e. identify the `emitc.expression` that sets the condition variable and push it down `do {} while (/*HERE*/)` or push it up `while (/*HERE*/) {}` if possible, removing the condition variable. WDYT @kchibisov?

kchibisov replied:
> removing the condition variable. WDYT @kchibisov?

Yeah, that's what I've suggested. Like first translate to something like

```
bool cond = true;
do {
  cond = <SCF.COND>;
  if (cond) {
    <SCF.BODY>
  }
} while (cond)
```

(well, with variations), as it does now IIRC. And then just add a pass (if desired) to optimize this into something more readable. This pattern is easily identified, so wouldn't worry much about it.

The good thing that this pattern will always translate.

Variation with `while` is also possible, but semantically, it's a `do` loop, since you always have it started with `true`.

Vladislave0-0 committed:
[mlir][emitc] Changes after review

Vladislave0-0 committed:
[mlir][emitc] Changes after review

Deleted lovering from scf.while to emitc.while. Minor code style
changes have been made.

Vladislave0-0 committed:
[mlir][emitc] Changes after updating ExpressionOp

Vladislave0-0 committed:
[mlir][emitc] Update description (NFC)

aniragil commented:
> @aniragil, gentle ping. I've corrected the code based on the discussion above — removed lovering from `scf.while` to `emitc.while`.

Thanks for the ping!
As (I think) we discussed in the comments, the patch now introduces a dead `emitc.while` op, which it shouldn't. It should be commited along with the pass that transforms `emitc.do` into `emitc.while` as discussed. I'd do that it in a separate patch to first get the canonical scf lowering in (but if you prefer to add it here I guess that's fine too).

aniragil commented:
> @aniragil to move forward with this patch in a bit faster pace, can we then remove `emitc.while` for a while keeping only `emitc.do`, and process only with `do` for the time being. And then add the `emitc.while` in a separate patch with the pass that refactors it?

Exactly.

> 
> I think that should be easier to review then?

Yes, and it also separates the lowering from the optimization such that if the second patch gets reverted for any reason (e.g. bug) it won't revert the lowering part with it.

Vladislave0-0 committed:
[mlir][emitc] Remove unused WhileOp implementation

aniragil commented on mlir/lib/Dialect/EmitC/IR/EmitC.cpp:
```diff
@@ -974,10 +974,10 @@ LogicalResult emitc::YieldOp::verify() {
   Operation *containingOp = getOperation()->getParentOp();
 
-  if (result && containingOp->getNumResults() != 1)
+  if (result && containingOp->getNumResults() != 1 && !isa<DoOp>(containingOp))
```
More accurately can only be the condition region, right? (also below)
Better to first check concretely for `DoOp` and else make the generic check?

Vladislave0-0 committed:
[mlir][emitc] Changes after review

aniragil commented on mlir/lib/Conversion/SCFToEmitC/SCFToEmitC.cpp:
```diff
@@ -336,11 +336,205 @@ LogicalResult IndexSwitchOpLowering::matchAndRewrite(
+                              ConversionPatternRewriter &rewriter,
+                              Location loc) const {
+    auto yieldOp = cast<scf::YieldOp>(terminator);
+    SmallVector<Value> yields(yieldOp.getOperands());
```
Yield handling in other ops (among other things in this file) seems to have been fixed in https://github.com/llvm/llvm-project/pull/121476, @mgehre-amd could you take a look at this patch w.r.t. those changes you made there?

Vladislave0-0 committed:
[mlir][emitc] Changes after review

aniragil commented on mlir/lib/Conversion/SCFToEmitC/SCFToEmitC.cpp:
```diff
@@ -336,11 +337,214 @@ LogicalResult IndexSwitchOpLowering::matchAndRewrite(
+    rewriter.inlineRegionBefore(whileOp.getAfter(), ifOp.getBodyRegion(),
+                                ifOp.getBodyRegion().begin());
+
+    if (!ifOp.getBodyRegion().empty()) {
```
If the `if` is empty, let's not create it.

aniragil replied:
> If the `if` is empty, let's not create it.

Would be good to add such a test and the opposite test (only scf.condition in the before region) for coverage.

Vladislave0-0 committed:
[mlir][emitc] Do not create an ifOp if the after-region is empty

Vladislave0-0 committed:
[mlir][emitc] Duplicate of the 'lowerYield' function has been removed

Vladislave0-0 commented on mlir/lib/Conversion/SCFToEmitC/SCFToEmitC.cpp:
```diff
@@ -336,11 +337,214 @@ LogicalResult IndexSwitchOpLowering::matchAndRewrite(
+
+  // Convert SCF yield terminators to imperative assignments to update loop
+  // variables, maintaining loop semantics while transitioning to emitc model.
+  LogicalResult processYieldTerminator(Operation *terminator,
```
Thanks, I removed it.

Vladislave0-0 committed:
[mlir][emitc] Changes after review

Vladislave0-0 committed:
[mlir][emitc] Fixed a bug with return values in the lowering

Vladislave0-0 commented on mlir/test/Conversion/SCFToEmitC/while.mlir:
```diff
@@ -0,0 +1,238 @@
+  return %result : i32
+}
+
+func.func @one_result() -> i32 {
```
Test:
```mlir
func.func @one_result() -> i32 {
  %init = emitc.literal "1.0" : i32
  %var  = emitc.literal "1.0" : i32
  %exit = emitc.literal "10.0" : i32

  %res = scf.while (%arg1 = %init) : (i32) -> i32 {
    %sum = emitc.add %arg1, %var : (i32, i32) -> i32
    %condition = emitc.cmp lt, %sum, %exit : (i32, i32) -> i1
    %next = emitc.add %arg1, %arg1 : (i32, i32) -> i32
    scf.condition(%condition) %next : i32
  } do {
  ^bb0(%arg2: i32):
    %next_arg1 = emitc.call @payload_one_result(%arg2) : (i32) -> i32
    scf.yield %next_arg1 : i32
  }
  
  return %res : i32
}
```

At the moment, lowering looks like this:
```mlir
func.func @one_result() -> i32 {
  %0 = emitc.literal "1.0" : i32
  %1 = emitc.literal "1.0" : i32
  %2 = emitc.literal "10.0" : i32
  %3 = "emitc.variable"() <{value = #emitc.opaque<"">}> : () -> !emitc.lvalue<i32> // results
  %4 = "emitc.variable"() <{value = #emitc.opaque<"">}> : () -> !emitc.lvalue<i32> // arguments
  emitc.assign %0 : i32 to %4 : <i32>
  %5 = "emitc.variable"() <{value = #emitc.opaque<"">}> : () -> !emitc.lvalue<i1>   // condition
  emitc.do {
    %7 = load %4 : <i32>
    %8 = add %7, %1 : (i32, i32) -> i32
    %9 = cmp lt, %8, %2 : (i32, i32) -> i1
    %10 = add %7, %7 : (i32, i32) -> i32
    assign %10 : i32 to %3 : <i32>    // new results
    assign %9 : i1 to %5 : <i1>       // new condition
    if %9 {
      %11 = call @payload_one_result(%10) : (i32) -> i32
      assign %11 : i32 to %4 : <i32>  // new arguments
    }
  } while {
    %7 = expression %5 : (!emitc.lvalue<i1>) -> i1 {
      %8 = load %5 : <i1>
      yield %8 : i1
    }
    yield %7 : i1
  }
  %6 = emitc.load %3 : <i32>
  return %6 : i32
}
```

aniragil commented on mlir/lib/Dialect/EmitC/IR/EmitC.cpp:
```diff
@@ -1561,6 +1561,82 @@ LogicalResult GetFieldOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+                     "'emitc.expression' from this condition region");
+
+  Block &bodyBlock = getBodyRegion().front();
+  if (bodyBlock.empty())
+    return emitOpError("body region cannot be empty");
```
Why is it illegal to have an empty body?
A `do {} while (cond);` loop is still valid C/C++, right? (and may be useful for implementing busy-wait with a volatile/side-effect based `cond`)

Vladislave0-0 replied:
Yes, I agree, thank you.

Vladislave0-0 committed:
[mlir][emitc] Added the ability to do empty bodyBlock

aniragil reviewed (changes requested):
Final nits.

aniragil commented on mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:
```diff
@@ -1727,4 +1727,107 @@ def EmitC_GetFieldOp
+  let hasVerifier = 1;
+
+  let extraClassDeclaration = [{
+    Operation *getRootOp();
```
That's the function's definition, but is it used anywhere? I couldn't find any calls to it in the patch.

Vladislave0-0 committed:
[mlir][emitc] Changes after review

Deleted unused function DoOp::getRootOp(). The tests have been
rewritten with symbolic variable names.

aniragil commented on mlir/test/Target/Cpp/do.mlir:
```diff
@@ -0,0 +1,164 @@
+// RUN: mlir-translate -mlir-to-cpp %s | FileCheck --match-full-lines %s -check-prefix=CPP-DEFAULT
+
+
+// CPP-DEFAULT-LABEL: void emitc_do(int32_t* v1) {
```
> By the way, there are several files that are written without symbolic variable names.

Yes, some must have slipped through reviews :(

> Would it make sense to fix them?

Yes (not in this patch of course), as it makes tests more robust against naming scheme changes.

aniragil reviewed (approved):
LGTM, thanks!

Vladislave0-0 commented:
> > LGTM, thanks!
> 
> @Vladislave0-0 do you need me to merge the PR?

Yes, that would be nice, since I don't have merge rights. Thanks!
